Script to generate Material Sheet options

nh916 opened the pull request:
# Script to generate `Material Sheet` controlled vocabulary

## Description

This is the implementation that generates all the options for the Excel Uploader material sheet.

This code is only for the material sheet because the other sheets do not have nesting and can just be copied and pasted from the online controlled vocabulary.

## How it works
It works by:
1. Getting the controlled vocabulary from CRIPT by copying each part and pasting it into the `source.xlsx` sheets
2. `material_sheet_keys.py` generates all the options by:
   1. reading each sheet of the `source.xlsx` file
   2. converting each sheet into a df
   3. running an algorithm to generate all the controlled vocabulary options up to 2 levels deep eg `property:condition`
   > Note: It does not include any ids for the options eg `[1]property:condition`
3. After creating all the options, then it outputs it into a new file called `utils/excel_files/output.xlsx`
4. The developer can then copy the options generated in the `output.xlsx` and paste them into the hidden sheet of the `CRIPT_template.xlsx`

nh916 committed:
Merge remote-tracking branch 'origin/master'

nh916 committed:
Merge branch 'master' of https://github.com/C-Accel-CRIPT/cript-excel-uploader

nh916 committed:
fixed documentation for data sheet sample_preparation

nh916 committed:
added new code design for single options and nested options and getting preferred units function

nh916 committed:
without nesting options can populate a df successfully

get_preferred_unit method works

single_options can return a successful df

nh916 committed:
started working on the double for loop for nesting

nh916 committed:
double for loop approach works so far with some extra commented out code

nh916 committed:
removed some commented out code from the double for loop

nh916 committed:
code clean up and source.xlsx is exactly the same

nh916 committed:
removed unused import

nh916 committed:
write_to_excel is working correctly

nh916 committed:
write_to_excel is working correctly

nh916 committed:
added identifiers to the list of options for material

added some comments and doc strings for clarity

nh916 committed:
created get instructions function

nh916 committed:
sorting the values before writing them

nh916 committed:
added code for property:attribute and property:relation

nh916 committed:
removed destination.xlsx because it is not needed and is generated instead and called output.xlsx

nh916 committed:
added attributes and relation sheet to source.xlsx

nh916 committed:
formatted code with black

nh916 committed:
renamed variables from row_1 to category and row_2 to column_name

nh916 committed:
added instructions for "property:relation" column

nh916 committed:
formatted material_sheet_keys.py with black

nh916 committed:
removed sample_preparation from attributes list

nh916 committed:
added instructions for uncertainty

nh916 committed:
code clean up: added comments, formatted code, removed unneeded variables, changed instruction for uncertainty_type

nh916 committed:
renamed function to make it more intuitive

nh916 committed:
changes to source.xlsx

nh916 committed:
sorting options by Name column to make it easier for users to find what they need from the dropdown

nh916 committed:
changed CAS to chemical_id and removed the description of it

nh916 committed:
updated material with the newest keys

nh916 committed:
updated material with the newest keys

nh916 committed:
added label/links to source.xlsx

the label/link makes it easier as to where to get the newest controlled vocabulary, and skipping the first row with Pandas to skip the first label/link

nh916 committed:
created short documentation called README.md within utils/

This documentation aims to make anyone reading the code understand the purpose, how it works, and how to use it

nh916 committed:
Merge branch 'develop' into develop_excel_file_options

CVilla17 reviewed (commented):
A while back I learned that it's not great practice to keep an excel file in your master branch of code if you're using github because no matter when you clone the repo that file is downloaded in some regard. For example, if later on we got rid of that file or replace it with a different one, you would still have to download the old file as part of the repo's history. The code is great and  helpful, but I would suggest keeping the file in the organizations drive and calling on it locally whenever we want to run this.

CVilla17 commented on utils/material_sheet_keys.py:
```diff
@@ -0,0 +1,249 @@
+    unit = "unit"
+    instructions = "instructions"
+
+    df = pd.DataFrame(columns=[category, column_name, unit, instructions])
```
Creating variables for these names is a bit redundant

nh916 replied:
I thought those would improve code readability and if we ever needed to change things it would be a bit easier, but I think you're right and it's not needed.

nh916 committed:
code clean up in the return of the get_new_df()

nh916 committed:
code clean up inside of get_new_df() by removing unneeded variables

nh916 committed:
adding property:method

nh916 committed:
updated methods instructions

nh916 committed:
changed source.xlsx to material_source.xlsx

nh916 committed:
renamed output to material_output.xlsx

because other sheets will need this algorithm too and its good to differentiate

nh916 committed:
updated material_source.xlsx

nh916 committed:
formatted the code

brili commented:
@nh916 does this still include a manual step of copy-paste the controlled vocab? we should aim to eliminate the manual process

nh916 committed:
added process options

nh916 committed:
fixed process vocab to have correct options

fixed process_source.xlsx and material_source.xlsx to have correct vocabulary

nh916 committed:
renamed from material_sheet_keys to shee_options_generator

renamed because it is for both process sheet and material sheet

nh916 commented:
@brili it does. If I can make the request to an endpoint to get a JSON of keys I think that could help eliminate this issue

nh916 commented:
@CVilla17 hmmm, I dont think that would happen from my experience but if you have an example or something let me know and I can look more into it

CVilla17 commented:
> @CVilla17 hmmm, I dont think that would happen from my experience but if you have an example or something let me know and I can look more into it

From what I've read online people suggest using something called Git LFS for things like excel files, especially if they are over 100MB

nh916 commented:
@CVilla17 yeah I've read about that too, but I think that is only for Large File Storage because Git will reject pushes that are more than 100MB. I think the best test for this is to create a separate repo and try to insert an Excel file and then remove it and see if it still happens

nh916 committed:
added update existing option

nh916 committed:
fixed typo

nh916 commented:
@bearmit do you have the endpoints for how to get the controlled vocabulary?

nh916 committed:
computational process autofill working

nh916 committed:
software_configuration options working

nh916 committed:
formatted sheet_options_generator.py

nh916 committed:
latest update before switching branches
